asset-transfer-basic/application-gateway-java/src/main/java/application/java/App.java

```diff
@@ -0,0 +1,288 @@
+	 * while waiting for the commit notification.
+	 */
+	private static void transferAssetAsync(Contract contract) throws GatewayException, CommitException {
+
+		// Submit a transaction, blocking until the transaction has been committed on
+		// the ledger.
+		System.out.println("\n" + "--> Async Submit Transaction: TransferAsset, updates existing asset owner");
+		SubmittedTransaction commit = contract.newProposal("TransferAsset")
+				.addArguments(assetId, "Saptha").build().endorse().submitAsync();
+         
+		
+		byte[] result = commit.getResult();
+		String oldOwner = new String(result, StandardCharsets.UTF_8);
+	    
+		// Update UI or reply to REST request before waiting for commit status
+		System.out.println("*** Successfully submitted transaction to transfer ownership from " +  oldOwner + " to Saptha");
+		System.out.println("*** Waiting for transaction commit");
+		Status status = commit.getStatus();
+		if (!status.isSuccessful()) {
+			// Commit failure
+			throw new Error("Transaction "+ status.getTransactionId() + " failed to commit with status code " + status.getCode());
+			
+		}
+		System.out.println("*** Transaction committed successfully");
+
+	}
+
+	private static void readAssetById(Contract contract) throws GatewayException, CommitException {
+
+		System.out.println("\n" + "--> Evaluate Transaction: ReadAsset, function returns asset attributes");
+		
+		// Submit a transaction, blocking until the transaction has been committed on
+		// the ledger.
+
+
+		byte[] evaluateResult = contract.evaluateTransaction("ReadAsset", assetId);
+		System.out.println("*** Result:" + prettyJson(evaluateResult));
+	}
+	
+	/**
+	 * submitTransaction() will throw an error containing details of any error responses from the smart contract.
+	 */
+	private static void updateNonExistentAsset(Contract contract) throws GatewayException, CommitException {
+
+		try {
+		// Submit a transaction, blocking until the transaction has been committed on
+		// the ledger.
+		System.out.println("\n" + "--> Submit Transaction: UpdateAsset asset70, asset70 does not exist and should return an error");
+		contract.submitTransaction("UpdateAsset", "asset70", "blue", "5", "Tomoko", "300");
+		System.out.println("******** FAILED to return an error");
+		
+		} catch (EndorseException | SubmitException | CommitStatusException e) {
+            // Any error that originates from a peer or orderer node external to the gateway will have its details
+            // embedded within the gRPC status error. The following code shows how to extract that.
+			System.out.println("*** Successfully caught the error: "+ "\n" + e);
+			e.printStackTrace();
+			for (ErrorDetail detail : e.getDetails()) {
+            	System.out.println("\n"+ "Error from endpoint: " + "\n" + "address: " + detail.getAddress() + "\n" +
+                        "mspId: " + detail.getMspId() + "\n" +"message:" + detail.getMessage() + "\n");
```

Comment:
I would suggest moving the "Error from endpoint" message outside the for loop so it is only printed once, followed by all the detail messages. You could even only output the message if you actually have details by checking !e.getDetails().isEmpty().

I would also suggest calling it "Details" (or "Error details") since "details" are mentioned in the error messages and the tutorial text.

Reply:
I would be tempted to use ", " as a separator instead of "\n" so each detail message is printed on a single line, but your choice

Reply:
changes done